TERMS.md

```diff
@@ -0,0 +1,97 @@
+- A Property is defined as a single item but the data type could be an arbitrary data type, including composite data types.
+- The singular nature of a property suggests that its value shall updated and communicated as an atomic operation, i.e. it is either fully updated or not at all.
+
+- The method of update and communication is typically target-environment specific, but the most common expectation is that a Property can be updated and _observed_, meaning that a client can call a method to get or set the current value of the property (with appropriate access control rules) and often also _subscribe_ to be notified of when the value of a property was changed for any reason.
+- In some other language definitions this type of item is called an **Attribute** or **Field**.
+- A property can also be seen as analogous (and bi-directionally translatable, if datatypes allow it) to a Sensor Signal in the Vehicle Signal Specification (VSS).
+
+**Side proposal**: Consider renaming Property to Attribute to match Franca IDL, and because properties have another use in OpenAPI and AsyncAPI, and it seems other HTTP-related communication as well.
+
+
+## Return values
+
+- A Return Value specifies what is returned from a Method execution.  Although "out parameters" could serve a similar purpose, the Return Value has been given special status.
+
+- There is a single return value type, but it can use any data type, including composite types.
+
+- In computing systems there are two main paradigms regarding invocation of a procedure/function: Blocking/synchronous vs. Non-blocking/asynchronous.
+
+  - In a blocking/synchronous environment, the client that invokes an operations (calls a function) will stop until the operation is completed and "returns".  Upon return, a return-value (result) is communicated.
+  - In an asynchronous setup, a client can trigger the start of an operation and then continue doing other work while it executes in parallel.  In this case, the executing function will report back when the operation is finished, but it could also give continuous reports as to the progress of the function ("streaming updates").
+
+- In the VSC-language, methods can be defined without defining their invocation strategy (synchronous/asynchronous), which is instead often defined at a later time, and may be constrained by the target environment capability.
+
+- In all cases, a Method in VSC-lang (optionally) defines a Return Value type.  The type could frequently be an enumeration type to communicate one of a predefined set of results, but it is free to be any datatype.
+
+- Although details are target-dependent, this is the expected behavior of the return type:
+  - In the case of a synchronous/blocking call, this value type is the return value of the function that is communicated (once) after the operation completes.
+  - In the case of an asynchronous environment, the return value type is also used, with the difference that it can be returned _multiple times_.
```

Comment:
I am not fully convinced here. I think in many cases the business logic on server side for a method could be quite different depending on it if it intends to return a stream of data/information or return only once.

Reply:
- Examples would be good.
- Also, if you can propose a "clean" way to specify this as two different return types in the language?

...  I wrote this because I believe it's the most common and best approach for the _majority_ of cases is to define the functional APIs independent of the way it is actually called.  

Then, perhaps layers can be used to redefine the return type, for the uncommon cases?

I mean.... in the end this is also target-environment dependent isn't it?  Some environments might not even have the ability to do asychronous calls.

Some programming environments also kind of work this way, i.e. they allow you to spawn off a thread to run a function (asynchronous), or to run it in the same thread (blocking) but that doesn't mean the function signature changes.  This is what leads me to believe that we should design _primarily_ for a function definition that is _independent_ of asynch/sync.

Reply:
> I am not fully convinced here. I think in many cases the business logic on server side for a method could be quite different depending on it if it intends to return a stream of data/information or return only once.

I thought about this some more and for this (and similar topics) could it be an OK answer to say:
You can always design your interfaces in a specific way with specific things in mind, even if the capabilities of the system is designed to be more general.  So... if you now you will always use certain interfaces in asynchronous mode, then I suppose you could design with that in mind.

In either case, I still like this solution because it enables both the ability to design interfaces without deciding if it's sync/async (if/when that makes sense), and from my perspective the return value is quite similar to a "status update".  I think it should definitely be possible to layer your status updates as well as your "result of the whole operation" into one return type.   (but also note, through layers you could redefine/extend whatever the return type was originally defined to be).